nickel/src/lib.rs

```diff
@@ -0,0 +1,637 @@
+    pub fn is_bool(&self) -> bool {
+        self.as_bool().is_some()
+    }
+
+    /// Is this expression a number?
+    pub fn is_num(&self) -> bool {
+        matches!(self.rt.as_ref(), Term::Num(_))
+    }
+
+    /// Is this expression a number?
+    pub fn as_num(&self) -> Option<Number<'_>> {
+        match self.rt.as_ref() {
+            Term::Num(num) => Some(Number { num }),
+            _ => None,
+        }
+    }
+
+    /// If this expression is an integer within the range of an `i64`, return it.
+    pub fn as_i64(&self) -> Option<i64> {
+        self.as_num().and_then(|n| n.as_i64())
+    }
+
+    /// If this expression is a number, round it to an `f64` and return it.
+    pub fn as_f64(&self) -> Option<f64> {
+        self.as_num().map(|n| n.as_f64())
+    }
+
+    /// If this expression is a string, return it.
+    pub fn as_str(&self) -> Option<&str> {
+        match self.rt.as_ref() {
+            Term::Str(s) => Some(s),
+            _ => None,
+        }
+    }
+
+    /// Is this expression a string?
+    pub fn is_str(&self) -> bool {
+        self.as_str().is_some()
+    }
+
+    /// If this expression is an enum tag, return it.
+    pub fn as_enum_tag(&self) -> Option<&str> {
+        match self.rt.as_ref() {
+            Term::Enum(tag) => Some(tag.label()),
+            _ => None,
+        }
+    }
+
+    /// Is this expression an enum tag?
+    pub fn is_enum_tag(&self) -> bool {
+        self.as_enum_tag().is_some()
+    }
+
+    /// If this expression is an enum variant, return its tag and its payload.
+    pub fn as_enum_variant(&self) -> Option<(&str, Expr)> {
+        match self.rt.as_ref() {
+            Term::EnumVariant { tag, arg, attrs: _ } => {
+                Some((tag.label(), Expr { rt: arg.clone() }))
+            }
+            _ => None,
+        }
+    }
+
+    /// Is this expression an enum variant?
+    pub fn is_enum_variant(&self) -> bool {
+        matches!(self.rt.as_ref(), Term::EnumVariant { .. })
+    }
+
+    /// Is this expression a record?
+    pub fn is_record(&self) -> bool {
+        self.as_record().is_some()
+    }
+
+    /// If this expression is a record, return a handle to it.
+    ///
+    /// The returned record handle can be used to access the
+    /// record's elements.
+    pub fn as_record(&self) -> Option<Record<'_>> {
+        match self.rt.as_ref() {
+            Term::Record(data) => Some(Record { data }),
+            _ => None,
+        }
+    }
+
+    /// If this expression is an array, return a handle to it.
+    ///
+    /// The returned array handle can be used to access the
+    /// array's elements.
+    pub fn as_array(&self) -> Option<Array<'_>> {
+        match self.rt.as_ref() {
+            Term::Array(array, _) => Some(Array { array }),
+            _ => None,
+        }
+    }
+
+    /// Is this expression an array?
+    pub fn is_array(&self) -> bool {
+        self.as_array().is_some()
+    }
+
+    pub fn is_thunk(&self) -> bool {
```

Comment:
I think the naming here is confusing. Internally we use "thunk" to mean something else; for example a bare "1 + 1" is not a thunk at runtime but is indeed not an effective weak head normal form.

Maybe exposing `is_whnf` could be closer to the Nickel model, with the same definition (still using the `is_eff_whnf`). In fact the latter (`is_eff_whnf`) is a bit of an artifact of the current representation, but it will be gone with #2302, and there will only be one notion of WHNF, so we might get away with not exposing the `eff_whnf/whnf` subtlety to the user at all.